hera_qm/firstcal_metrics.py

```diff
@@ -0,0 +1,542 @@
+        if output == True:
+            return self.metrics
+
+        """
+        # Calculate std and zscores
+        (self.ant_avg, self.ant_std, self.time_std, self.agg_std,
+         self.z_scores) = self.delay_std(return_dict=True)
+
+        # Given delay std cut find "bad" ant
+        # determine if full sol is bad
+        if self.agg_std > std_cut:
+            self.full_sol = 'bad'
+        else:
+            self.full_sol = 'good'
+
+        self.bad_ants = []
+        for ant in self.ant_std:
+            if self.ant_std[ant] > std_cut:
+                self.bad_ants.append(ant)
+
+        # put into dictionary
+        metrics = OrderedDict()
+        metrics['full_sol'] = self.full_sol
+        metrics['bad_ant'] = self.bad_ants
+        metrics['z_scores'] = self.z_scores
+        metrics['ant_avg'] = self.ant_avg
+        metrics['ant_std'] = self.ant_std
+        metrics['time_std'] = self.time_std
+        metrics['agg_std'] = self.agg_std
+        metrics['times'] = self.times
+        metrics['version'] = self.version_str
+        self.metrics = metrics
+
+        if output is True:
+            return self.metrics
+
+    def write_metrics(self, filename=None, filetype='json'):
+        """
+        Write metrics to file after running run_metrics()
+
+        filename : str, default=None
+            filename without filetype suffix
+            if None, use default filename stem
+
+        filetype : str, default='json'
+            filetype to write out to
+            options = ['json', 'pkl']
+
+        """
+        # get filename prefix
+        if filename is None:
+            filename = self.file_stem + ".metrics"
+
+        # write to file
+        if filetype == 'json':
+            if filename.split('.')[-1] != 'json':
+                filename += '.json'
+            # change ndarrays to lists
+            metrics_out = copy.deepcopy(self.metrics)
+            metrics_out['times'] = list(metrics_out['times'])
+            for k in metrics_out['z_scores'].keys():
+                metrics_out['z_scores'][k] = list(metrics_out['z_scores'][k])
+
+            with open(filename, 'w') as f:
+                json.dump(metrics_out, f, indent=4)
+
+        elif filetype == 'pkl':
+            if filename.split('.')[-1] != 'pkl':
+                filename += '.pkl'
+            with open(filename, 'wb') as f:
+                outp = pkl.Pickler(f)
+                outp.dump(self.metrics)
+
+    def load_metrics(self, filename):
+        """
+        Read-in a firstcal_metrics file and append to class
+
+        Input:
+        ------
+        filename : str
+            filename to read in
+
+        """
+        # get filetype
+        filetype = filename.split('.')[-1]
+        if filetype == 'json':
+            with open(filename, 'r') as f:
+                self.metrics = json.load(f)
+
+            # ensure keys of dicts are not strings
+            for i in self.metrics['ant_avg'].keys():
```

Comment:
I spent too much time yesterday trying to get this to work. ant_metrics uses tuples as keys, which json doesn't like. As far as I can tell, the best thing to do is convert them to strings, but then eval is necessary. All this to say we should revisit when we're unifying the two classes.